engine/engine.go

```diff
@@ -50,82 +64,134 @@ func New(ctx context.Context, cliCtx *cli.Context) (*Engine, error) {
 			continue
 		}
+		return task // found a runable task
+	}
+}
 
-		log.Infow("successfully acquired task", "uuid", task.UUID)
+func (e *Engine) worker(ctx context.Context, n int) {
+	defer e.wg.Done()
 
-		finalStatus := tasks.Successful
-		updateStage := func(stage string, stageDetails tasks.StageDetails) error {
-			task, err = e.client.UpdateTask(ctx, task.UUID.String(),
-				tasks.Type.UpdateTask.OfStage(
-					e.host,
-					tasks.InProgress,
-					stage,
-					stageDetails,
-				))
-			return err
-		}
-		if task.RetrievalTask.Exists() {
-			ctx := context.TODO()
-			err = tasks.MakeRetrievalDeal(ctx, e.nodeConfig, e.node, task.RetrievalTask.Must(), updateStage, log.Infow)
-			if err != nil {
-				finalStatus = tasks.Failed
-				log.Errorw("retrieval task returned error", "err", err)
-			} else {
-				log.Info("successfully retrieved data")
-			}
-		}
+	log.Infow("engine worker started", "worker_id", n)
 
-		if task.StorageTask.Exists() {
-			ctx := context.TODO()
-			err = tasks.MakeStorageDeal(ctx, e.nodeConfig, e.node, task.StorageTask.Must(), updateStage, log.Infow)
-			if err != nil {
-				finalStatus = tasks.Failed
-				log.Errorw("storage task returned error", "err", err)
-			} else {
-				log.Info("successfully stored data")
+	for {
+		// Check if there is a new task
+		task := e.popTask(ctx)
+		if task == nil {
+			// No tasks to run, so wait
+			select {
+			case <-ctx.Done():
+				return // engine closed, worker exits.
+			case <-time.After(noTasksWait):
 			}
+			continue
 		}
-		_, err = e.client.UpdateTask(ctx, task.UUID.String(),
+
+		log.Infow("successfully acquired task", "uuid", task.UUID)
+		e.runTask(ctx, task)
+	}
+}
+
+func (e *Engine) runTask(ctx context.Context, task tasks.Task) {
+	var err error
+
+	// Define function to update task stage.  Use engine context, not tasks context.
+	updateStage := func(stage string, stageDetails tasks.StageDetails) error {
+		task, err = e.client.UpdateTask(ctx, task.UUID.String(),
 			tasks.Type.UpdateTask.OfStage(
 				e.host,
-				finalStatus,
-				task.Stage.String(),
-				task.CurrentStageDetails.Must(),
+				tasks.InProgress,
+				stage,
+				stageDetails,
 			))
+		return err
+	}
+
+	// Create a context to manage the lifetime of the current task
+	taskCtx, taskCancel := context.WithTimeout(ctx, maxTaskLifetime)
+	defer taskCancel()
+
+	finalStatus := tasks.Successful
+
+	// Start deals
+	if task.RetrievalTask.Exists() {
+		err = tasks.MakeRetrievalDeal(taskCtx, e.nodeConfig, e.node, task.RetrievalTask.Must(), updateStage, log.Infow)
 		if err != nil {
-			log.Error("Error updating final status")
+			if err == context.Canceled {
+				// Engine closed, do not update final state
+				log.Warn("task", task.UUID.String(), "canceled for shutdown")
+				return
+			}
+			finalStatus = tasks.Failed
+			log.Errorw("retrieval task returned error", "err", err)
+		} else {
+			log.Info("successfully retrieved data")
+		}
+	}
+	if task.StorageTask.Exists() {
```

Comment:
else if? or error if both exist

Reply:
Copied from original logic in engine, so I thought this may have been in case some task could have both types.
https://github.com/filecoin-project/dealbot/blob/main/engine/engine.go#L110

Changing to `else if`